common/src/main/java/io/pravega/common/concurrent/ThreadPoolScheduledExecutorService.java

```diff
@@ -0,0 +1,376 @@
+                handler);
+        runner.prestartAllCoreThreads();
+    }
+
+    @RequiredArgsConstructor
+    private final class CancelableFuture<R> implements ScheduledFuture<R> {
+
+        private final ScheduledRunnable<R> task;
+
+        @Override
+        public long getDelay(TimeUnit unit) {
+            if (!task.isDelayed) {
+                return 0;
+            }
+            return unit.convert(task.scheduledTimeNanos - System.nanoTime(), TimeUnit.NANOSECONDS);
+        }
+
+        @Override
+        public int compareTo(Delayed other) {
+            if (other == this) { // compare zero if same object
+                return 0;
+            }
+            if (other instanceof CancelableFuture) {
+                return Long.compare(this.task.scheduledTimeNanos,
+                                    ((CancelableFuture<?>) other).task.scheduledTimeNanos);
+            } else {
+                long diff = getDelay(NANOSECONDS) - other.getDelay(NANOSECONDS);
+                return (diff < 0) ? -1 : (diff > 0) ? 1 : 0;
+            }
+        }
+
+        /**
+         * Cancels a pending task. Note: The {@code mayInterruptIfRunning} parameter is ignored (and
+         * assumed to be false) as it is unsupported.
+         * 
+         * @param mayInterruptIfRunning Ignored.
+         */
+        @Override
+        public boolean cancel(boolean mayInterruptIfRunning) {
+            return ThreadPoolScheduledExecutorService.this.cancel(task);
+        }
+
+        @Override
+        public boolean isCancelled() {
+            return task.future.isCancelled();
+        }
+
+        @Override
+        public boolean isDone() {
+            return task.future.isDone();
+        }
+
+        @Override
+        public R get() throws InterruptedException, ExecutionException {
+            return task.future.get();
+        }
+
+        @Override
+        public R get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
+            return task.future.get(timeout, unit);
+        }
+    }
+
+    @Data
+    private static final class ScheduledRunnable<R> implements Runnable, Scheduled {
+        private final long id;
+        private final boolean isDelayed;
+        private final long scheduledTimeNanos;
+        @EqualsAndHashCode.Exclude
+        private final Callable<R> task;
+        @EqualsAndHashCode.Exclude
+        private final CompletableFuture<R> future;
+
+        private ScheduledRunnable(Callable<R> task) {
+            this.id = COUNTER.incrementAndGet();
+            this.isDelayed = false;
+            this.scheduledTimeNanos = 0;
+            this.task = task;
+            this.future = new CompletableFuture<R>();
+        }
+
+        private ScheduledRunnable(Callable<R> task, long delay, TimeUnit unit) {
+            this.id = COUNTER.incrementAndGet();
+            this.isDelayed = true;
+            this.scheduledTimeNanos = unit.toNanos(delay) + System.nanoTime();
```

Comment:
I'm thinking that in the interest of testing, you might want to have a way to pass a time supplier (and default it to `System.nanoTime`). Since this will be core to all of our components, I want this properly tested and not rely on the actual passage of time to verify it (which will likely slow down the tests).

Reply:
While it technically relies on time, I did the unit tess carefully. The all of the tests together add up to 0.6 sec and they are robust against delays or other timing issues.